snuba/datasets/plans/partitioned_storage.py

```diff
@@ -0,0 +1,158 @@
+from abc import ABC, abstractmethod
+from typing import Optional, Sequence
+
+import sentry_sdk
+
+from snuba.clickhouse.query_dsl.accessors import get_object_ids_in_query_ast
+from snuba.clickhouse.translators.snuba.mapping import TranslationMappers
+from snuba.clusters.cluster import ClickhouseCluster, get_cluster
+from snuba.clusters.storage_sets import StorageSetKey
+from snuba.datasets.partitioning import (
+    is_storage_partitioned,
+    map_logical_partition_to_slice,
+    map_org_id_to_logical_partition,
+)
+from snuba.datasets.plans.query_plan import (
+    ClickhouseQueryPlan,
+    ClickhouseQueryPlanBuilder,
+)
+from snuba.datasets.plans.single_storage import (
+    SimpleQueryPlanExecutionStrategy,
+    get_query_data_source,
+)
+from snuba.datasets.plans.translator.query import QueryTranslator
+from snuba.datasets.storage import ReadableStorage
+from snuba.datasets.storages.storage_key import StorageKey
+from snuba.query.logical import Query as LogicalQuery
+from snuba.query.processors.physical import ClickhouseQueryProcessor
+from snuba.query.processors.physical.conditions_enforcer import (
+    MandatoryConditionEnforcer,
+)
+from snuba.query.processors.physical.mandatory_condition_applier import (
+    MandatoryConditionApplier,
+)
+from snuba.query.query_settings import QuerySettings
+from snuba.util import with_span
+
+
+class StorageClusterSelector(ABC):
+    """
+    The component provided by a dataset and used at the beginning of the
+    execution of a query to pick the storage set a query should be executed
+    onto.
+    """
+
+    @abstractmethod
+    def select_cluster(
+        self, query: LogicalQuery, query_settings: QuerySettings
+    ) -> ClickhouseCluster:
+        raise NotImplementedError
+
+
+class ColumnBasedStoragePartitionSelector(StorageClusterSelector):
+    """
+    Storage partition selector for the generic metrics storage. This is needed
+    because the generic metrics storage can be partitioned and we would need to
+    know which partition to use for a specific query.
+    """
+
+    def __init__(
+        self,
+        storage: StorageKey,
+        storage_set: StorageSetKey,
```

Comment:
these are both derivable from the `ReadableTableStorage` object itself but it seemed a little cleaner to pass the keys around -- happy to change that though